### ISSUE TITLE
Unable to Upload/Select Files from Copyediting stage

### ISSUE THREAD

**Issue by Baytan** [pkp/pkp-lib#7156]
in my case Copyediting it is not possible to use “Upload/Select Files” from Copyediting stage.

It happens that the window opens but does not load the part for uploading or selecting files.
![image](https://user-images.githubusercontent.com/27674815/123278869-ff563e80-d50f-11eb-9860-f5feed7f69a7.png)

**this is exactly the same case with https://github.com/pkp/pkp-lib/issues/4754 but it is not solved by the method there. The file revision is already done in our OJS 3.3.0-6 system. in other words, the given solution was already applied to the related files, but the problem still exists.
we are using php 7.4.**

the related error log was:
Stack trace:
#0 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(402): SubmissionFilesCategoryGridDataProvider->loadCategoryData(Object(Request), 4, Array)
#1 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(124): CategoryGridHandler->loadCategoryData(Object(Request), 4, Array)
#2 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(473): CategoryGridHandler->getGridCategoryDataElements(Object(Request), 4)
#3 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(453): CategoryGridHandler->_renderCategoryInternally(Object(Request), Object(SelectableSubmissionFileListCategoryGridRow))
#4 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc. in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/SubmissionFilesCategoryGridDataProvider.inc.php on line 131
[24-Jun-2021 16:59:33 Europe/Istanbul] PHP Warning:  Declaration of ManageCopyeditFilesForm::execute($stageSubmissionFiles, $fileStage = NULL) should be compatible with ManageSubmissionFilesForm::execute($stageSubmissionFiles = NULL, $fileStage = NULL, ...$functionArgs) in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/copyedit/form/ManageCopyeditFilesForm.inc.php on line 33
[24-Jun-2021 16:59:58 Europe/Istanbul] PHP Warning:  Declaration of ManageCopyeditFilesForm::execute($stageSubmissionFiles, $fileStage = NULL) should be compatible with ManageSubmissionFilesForm::execute($stageSubmissionFiles = NULL, $fileStage = NULL, ...$functionArgs) in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/copyedit/form/ManageCopyeditFilesForm.inc.php on line 33
[24-Jun-2021 16:59:59 Europe/Istanbul] PHP Fatal error:  Uncaught Error: Call to a member function getAssocId() on null in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/SubmissionFilesCategoryGridDataProvider.inc.php:131
Stack trace:

**the problem seems to be related with SubmissionFilesCategoryGridDataProvider.inc.php on line 131.
then without any knowledge I edited the code using two lines I found on internet**

**before code was like that:**
	$submissionFiles = $this->_submissionFiles;
			$stageSubmissionFiles = array();
			foreach ($submissionFiles as $key => $submissionFile) {
				if (in_array($submissionFile->getData('fileStage'), (array) $fileStage)) {
					$stageSubmissionFiles[$key] = $submissionFile;
				} elseif ($submissionFile->getData('fileStage') == SUBMISSION_FILE_QUERY) {
					// Determine the stage from the query.
					if ($submissionFile->getData('assocType')!=ASSOC_TYPE_NOTE) break;
					$noteDao = DAORegistry::getDAO('NoteDAO'); /* @var $noteDao NoteDAO */
					$note = $noteDao->getById($submissionFile->getData('assocId'));
					assert($note && $note->getAssocType()==ASSOC_TYPE_QUERY);
					$queryDao = DAORegistry::getDAO('QueryDAO'); /* @var $queryDao QueryDAO */
					$query = $queryDao->getById($note->getAssocId());
					if ($query && $query->getStageId() == $stageId) $stageSubmissionFiles[$key] = $submissionFile;
				}
			}
		}
		return $dataProvider->prepareSubmissionFileData($stageSubmissionFiles, false, $filter);

**after I add two lines: which are given bold**
	$submissionFiles = $this->_submissionFiles;
			$stageSubmissionFiles = array();
			foreach ($submissionFiles as $key => $submissionFile) {
				if (in_array($submissionFile->getData('fileStage'), (array) $fileStage)) {
					$stageSubmissionFiles[$key] = $submissionFile;
				} elseif ($submissionFile->getData('fileStage') == SUBMISSION_FILE_QUERY) {
					// Determine the stage from the query.
					if ($submissionFile->getData('assocType')!=ASSOC_TYPE_NOTE) break;
					$noteDao = DAORegistry::getDAO('NoteDAO'); /* @var $noteDao NoteDAO */
					$note = $noteDao->getById($submissionFile->getData('assocId'));
					assert($note && $note->getAssocType()==ASSOC_TYPE_QUERY);
					$queryDao = DAORegistry::getDAO('QueryDAO'); /* @var $queryDao QueryDAO */
					**if (!$submission) error_log('MISSING SUBMISSION ID ' . $submissionId . ': assoc_type ' . $notification->getAssocType() . ' and assoc_id ' . $notification->getAssocId() . '; type ' . $notification->getType());
					return $submission->getLocalizedTitle();**
					$query = $queryDao->getById($note->getAssocId());
					if ($query && $query->getStageId() == $stageId) $stageSubmissionFiles[$key] = $submissionFile;
				}
			}
		}
		return $dataProvider->prepareSubmissionFileData($stageSubmissionFiles, false, $filter);

then it worked and I was able to add files to the Copyedited part.
**But I do not know what I have done**
**Is it something bad and get more problems to me. How should I edit the code to be OK?**

after my code editing related error log is:
Stack trace:
#0 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(402): SubmissionFilesCategoryGridDataProvider->loadCategoryData(Object(Request), 4, Array)
#1 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(124): CategoryGridHandler->loadCategoryData(Object(Request), 4, Array)
#2 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(473): CategoryGridHandler->getGridCategoryDataElements(Object(Request), 4)
#3 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(453): CategoryGridHandler->_renderCategoryInternally(Object(Request), Object(SelectableSubmissionFileListCategoryGridRow))
#4 /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc. in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/SubmissionFilesCategoryGridDataProvider.inc.php on line 133
[24-Jun-2021 17:02:10 Europe/Istanbul] PHP Warning:  Declaration of ManageCopyeditFilesForm::execute($stageSubmissionFiles, $fileStage = NULL) should be compatible with ManageSubmissionFilesForm::execute($stageSubmissionFiles = NULL, $fileStage = NULL, ...$functionArgs) in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/copyedit/form/ManageCopyeditFilesForm.inc.php on line 33
[24-Jun-2021 17:02:10 Europe/Istanbul] PHP Warning:  Invalid argument supplied for foreach() in /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/GridHandler.inc.php on line 981
[24-Jun-2021 17:03:59 Europe/Istanbul] PHP Warning:  Declaration of ManageCopyeditFilesForm::execute($stageSubmissionFiles, $fileStage = NULL) should be compatible with ManageSubmissionFilesForm::execute($stageSubmissionFiles = NULL, $fileStage = NULL, ...$functionArgs) in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/copyedit/form/ManageCopyeditFilesForm.inc.php on line 33
[24-Jun-2021 17:04:00 Europe/Istanbul] PHP Warning:  Invalid argument supplied for foreach() in /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/GridHandler.inc.php on line 981
[24-Jun-2021 17:04:41 Europe/Istanbul] PHP Warning:  Invalid argument supplied for foreach() in /home/sappear1/public_html/sts/lib/pkp/classes/controllers/grid/GridHandler.inc.php on line 981
[24-Jun-2021 17:04:44 Europe/Istanbul] PHP Warning:  Declaration of ManageCopyeditFilesForm::execute($stageSubmissionFiles, $fileStage = NULL) should be compatible with ManageSubmissionFilesForm::execute($stageSubmissionFiles = NULL, $fileStage = NULL, ...$functionArgs) in /home/sappear1/public_html/sts/lib/pkp/controllers/grid/files/copyedit/form/ManageCopyeditFilesForm.inc.php on line 33
[24-Jun-2021 17:07:02 Europe/Istanbul] ojs2: 404 Not Found

**Comment by Baytan** [comment]
**I seem to solve the problem with code editing but I do not know what I have done. Please help me if I did something wrong and do more damage. Thanks to everybody in this OJS project and thanks in advance any body helps me.**

**Comment by Baytan** [comment]
This problem seems to be related with https://github.com/pkp/pkp-lib/issues/6774
All problematic submissions are listed in the query:
SELECT sf.submission_file_id FROM submission_files sf LEFT JOIN notes n ON (sf.assoc_id = n.note_id) WHERE sf.assoc_type = 520 AND n.note_id IS NULL;

I have deleted the related entries from the database, restore the code in SubmissionFilesCategoryGridDataProvider.inc.php to the original form, and everything seems to be fine.

I would like to know why my code revision worked anyway

**Comment by felixhelix** [comment]
I am experiencing the same problem: 
not able to upload files for copy editing.
@asmecher can you tag this as a bug so someone can have look at it?
your's
felix

**Comment by asmecher** [comment]
@felixhelix, which of our applications are you using and what version? Does your PHP error log contain messages from the time of failure?

**Comment by felixhelix** [comment]
Sorry for not being more specific:

We're using OJS 3.3.0.8
I have no access to the PHP error log in the moment, but from the browser console I get:

[HTTP/1.1 500 Internal Server Error 62ms]
https://journals.ub.uni-koeln.de/index.php/k_ON/$$$call$$$/grid/files/final/manage-final-draft-files-grid/fetch-grid?submissionId=852&_=1632242350143

Does this help?

your's
Felix

**Comment by asmecher** [comment]
Could you check the PHP error log, @felixhelix? That's the only way to know for sure if it's the same problem.

**Comment by felixhelix** [comment]
In the Apache error log I find the following related entries:
`
[Fri Oct 01 14:08:33.437616 2021] [php7:warn] [pid 11309] [client 134.95.147.149:40782] PHP Warning:  Declaration of ManageFinalDraftFilesForm::execute($stageSubmissionFiles, $fileStage = NULL) should be compatible with ManageSubmissionFilesForm::execute($stageSubmissionFiles = NULL, $fileStage = NULL, ...$functionArgs) in /var/www/html/lib/pkp/controllers/grid/files/final/form/ManageFinalDraftFilesForm.inc.php on line 18, referer: https://journals.xxx.de/index.php/xxx/workflow/index/852/4
`
`
[Fri Oct 01 14:08:33.544237 2021] [php7:error] [pid 13619] [client 134.95.147.149:40784] PHP Fatal error:  Uncaught Error: Call to a member function getAssocId() on null in /var/www/html/lib/pkp/controllers/grid/files/SubmissionFilesCategoryGridDataProvider.inc.php:131\nStack trace:\n#0 /var/www/html/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(402): SubmissionFilesCategoryGridDataProvider->loadCategoryData(Object(Request), 4, Array)\n#1 /var/www/html/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(124):CategoryGridHandler->loadCategoryData(Object(Request), 4, Array)\n#2 /var/www/html/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(473): CategoryGridHandler->getGridCategoryDataElements(Object(Request), 4)\n#3 /var/www/html/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(453): CategoryGridHandler->_renderCategoryInternally(Object(Request), Object(SelectableSubmissionFileListCategoryGridRow))\n#4 /var/www/html/lib/pkp/classes/controllers/grid/CategoryGridHandler.inc.php(276): CategoryGridHandler->_renderCategoriesInternally(Object(Request))\n#5 /var/www/html/lib/pkp/c in /var/www/html/lib/pkp/controllers/grid/files/SubmissionFilesCategoryGridDataProvider.inc.php on line 131, referer: https://journals.ub.xxx.de/index.php/xxx/workflow/index/852/4
`
Is this enough?

your's
Felix

**Comment by asmecher** [comment]
@felixhelix, yes, please do delete that entry -- but take a complete database backup first, and test afterwards!

**Comment by ohilbig01** [comment]
Hi @asmecher,
we have the same problem (can't upload files in copyediting stage), and the same messages in the php-error_log.
We are using OJS  3.3.0.7

I also did this select:
`SELECT sf.submission_id, sf.submission_file_id, sf.file_id FROM submission_files sf LEFT JOIN notes n ON (sf.assoc_id = n.note_id) WHERE sf.assoc_type = 520 AND n.note_id IS NULL `

![image](https://user-images.githubusercontent.com/25659317/140911609-c1b1d2a7-8280-43ce-8e67-4ab49bf116d2.png)


These are the files are in the files dir of submission 161:
```
ojs02/files/journals/1/articles # find 161
161
161/618163df323f2.docx
161/submission
161/submission/final
161/submission/final/161-1-418-1-6-20210707.docx
161/submission/final/161-1-327-1-6-20200531.doc
161/submission/query
161/submission/query/161-12-421-1-18-20210723.jpg
161/submission/query/161-1-435-1-18-20210723.docx
161/submission/query/161-12-433-1-18-20210723.jpg
161/submission/query/161-1-420-1-18-20210723.docx
161/submission/query/161-12-432-1-18-20210723.jpg
161/submission/query/161-12-423-1-18-20210723.jpg
161/submission/query/161-12-434-1-18-20210723.jpg
161/submission/query/161-12-422-1-18-20210723.jpg
161/submission/query/161-1-383-1-18-20210707.docx
161/submission/query/161-1-424-1-18-20210723.docx
161/submission/query/161-1-431-1-18-20210723.jpg
161/submission/query/161-1-420-2-18-20210723.jpg
161/submission/161-12-254-1-2-20200531.jpg
161/submission/161-12-251-1-2-20200531.jpg
161/submission/161-1-250-1-2-20200531.doc
161/submission/161-12-255-1-2-20200531.jpg
161/submission/161-12-252-1-2-20200531.jpg
```

I don't know if I should delete something in the database if there are these files in the folder?  After all, the file names seem to contain the file_id,
Regards

**Comment by asmecher** [comment]
@ohilbig01, I would recommend taking a good database backup, then removing those entries. They appear to be files related to notes that have since been deleted (or were never uploaded properly).

**Comment by ohilbig01** [comment]
Thank you @asmecher, do you mean this:
* remove the entries with the six selected submission_file_ids in the table submission_files
* delete the files that contain those submission_file_ids in their name?

**Comment by asmecher** [comment]
@ohilbig01, I mean to remove those six matching entries from from the `submission_files` table. There is no harm in leaving the files in place.

**Comment by ohilbig01** [comment]
Thanks @asmecher. I also had to delete references in submission_file_revisions and submission_file_settings.

**Comment by OhnO395** [comment]
https://github.com/pkp/pkp-lib/commit/35f6cd1350ce6f0e3dd1fb2a2a2dd4aadc825b63
fix draft upload only

**Comment by jnugent** [comment]
I'm not sure if this is related, but I am currently debugging a ticket by a hosted client. They have a copyeditor who is unable to upload revisions of copyedited files.  I'm still debugging this but what is happening is the PKPSubmissionFilesUploadBaseForm.inc.php class in the file upload wizard is returning "Invalid Parameters!" because the $stageId isn't making it into the constructor (I've verified that already). It *is* present in the POST request to upload the file however.

**Comment by jnugent** [comment]
Hi @asmecher 

I am not sure if this is the ultimate cause of this issue, but I worked through our problem and discovered that if you upload a revision of a file, as a copyeditor with no other role, there's a a missing WorkflowStageAccessPolicy add so there is no authorized object present.  I've fixed our problem by adding a class import and then a policy add to this block of code in the FileUploadWizardHandler.

```
                // Allow access to modify a specific file
                if ($submissionFileIdToValidate) {
                        $stageId = (int) $request->getUserVar('stageId');
                        import('lib.pkp.classes.security.authorization.SubmissionFileAccessPolicy');
                        $this->addPolicy(new SubmissionFileAccessPolicy($request, $args, $roleAssignments, SUBMISSION_FILE_ACCESS_MODIFY, $submissionFileIdToValidate));
                        import('lib.pkp.classes.security.authorization.WorkflowStageAccessPolicy');
                        $this->addPolicy(new WorkflowStageAccessPolicy($request, $args, $roleAssignments, 'submissionId', $stageId));
                // Allow uploading to review attachments
                }
```

It occurs to me that this is similar to the patch I submitted and @NateWr subsequently merged here:

https://github.com/pkp/pkp-lib/commit/1ebed56819f106803f1ae77157aaf50aa37d0f92

Which was for something similar.

**Comment by dunnleaddress** [comment]
> [35f6cd1](https://github.com/pkp/pkp-lib/commit/35f6cd1350ce6f0e3dd1fb2a2a2dd4aadc825b63) fix draft upload only

Hello @OhnO395 , 

We have experienced the same issue and I have applied this fix to my OJS version 3.3.0.8 - it seems to fix the issue (need more testing) - The Upload/Select Files dialogue is now successfully loaded for attachment. 

We always appreciate your and others' hard work to keep this software working smoothly.

Best Regards,

D.

**Comment by stutzmann72** [comment]
Hello
We've discovered this bug on a 3.3.0-10 installation. The only error message we got in the php error log was

`AH00690: no acceptable variant: /var/www/ojs/index`

Best
jan

**Comment by jnugent** [comment]
@asmecher  this issue can be closed now, the asset guard fix was merged into the 3.3.0.14 release.

**Comment by asmecher** [comment]
Thanks, @jnugent!
